### PR TITLE
os/bluestore: Prune deleted onodes

### PR DESCRIPTION
**Pull request by aclamk**
This PR is about cleanup of remains of onodes that are lingering in onode cache after deletion.
When deleting collections large amounts of nodes that are logically deleted `exists == false` eat up space, inflicting cache pressure on existing ones.
Now when all reference counters for onode except one in `onode_map` are relinquished, and onode has been deleted, it is removed from cache and from `onode_map`.

This PR supersedes #38423.

### CODE REVIEW

**Comment by tchaikov** [comment]
dropping from my batch as it conflicts with #37496

**Review by ifed01 — COMMENTED** [review]
Looks great!
IMO can be improved (in terms of fewer line numbers) a bit as per my comments.

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -1045,7 +1045,15 @@ struct LruOnodeCacheShard : public BlueStore::OnodeCacheShard {
     --num_pinned;
     dout(20) << __func__ << this << " " << " " << " " << o->oid << " unpinned" << dendl;
   }
-
+  void _rm_pinned(BlueStore::Onode* o) override
+  {
+    ceph_assert(!o->pinned);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:1050]: Given the function name this assertion looks a bit confusing. May be remove it totally similar to _unpin()? 
And rename to _unpin_and_rm()?
Not a big deal though...

**Review by ifed01 — APPROVED** [review]

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc]: @ifed01 Confusion is bad. I renamed it and deleted assert.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-12-16_15:58:13-rados-wip-kefu-testing-2020-12-16-1740-distro-basic-smithi
- https://pulpito.ceph.com/kchai-2020-12-17_12:24:37-rados-wip-kefu-testing-2020-12-17-1542-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-18_01:47:10-rados-wip-kefu-testing-2020-12-17-1542-distro-basic-smithi/

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-12-16_15:58:13-rados-wip-kefu-testing-2020-12-16-1740-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-17_12:24:37-rados-wip-kefu-testing-2020-12-17-1542-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-18_01:47:10-rados-wip-kefu-testing-2020-12-17-1542-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-19_06:27:52-rados-wip-kefu-testing-2020-12-17-1542-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-20_12:15:20-rados-wip-kefu-testing-2020-12-17-1542-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-21_03:37:05-rados-wip-kefu-testing-2020-12-17-1542-distro-basic-smithi/

i don't think the failures are related.

**Review by jdurgin — APPROVED** [review]
reviewed in the course of looking through #38651

**Pull request merged**